### ISSUE TITLE
[group conv backward] wrong value calculated

### ISSUE THREAD

**Issue by jay746** [pytorch/pytorch#50042]
## 🐛 Bug

group convolution backward path result is not correct.
This happens with specific combination of input only
if input and output channels are 24 and groups = 24
stride = 2 and padding = 1

As mentioned below if you change group_val to 25 or 23 , backward path output is correct. only with 24 it's gives wrong results.
Here grad_in is initialized to 1 hence output of backward path will be equal to forward path.

```
import torch
group_val = 24
ifm = torch.ones([1, group_val, 6, 6], dtype=torch.float32)
weights = torch.ones([group_val, 1, 3, 3], dtype=torch.float32)
op = torch.nn.Conv2d(
        in_channels=group_val,
        out_channels=group_val,
        kernel_size=[3,3],
        stride=[2,2],
        padding=[1,1],
        dilation=[1,1],
        groups=group_val,
        bias=False,
        padding_mode='zeros'
    )

op.weight.data = weights
res = op(ifm)
print(res)
grad_in = torch.ones(res.shape, dtype=torch.float32)
res.backward(grad_in)
print(op.weight.grad)
```

## Result (adding only 1 channel result as other's are same only):
tensor([[[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]],

## Expected behavior
Tensor([[[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]],

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): NA
 - Python version: 3.6.8
 - CUDA/cuDNN version: cpu
 - GPU models and configuration: na
 - Any other relevant information: na

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @mruberry

**Comment by ngimel** [comment]
High priority for silent wrong answer.

**Comment by gchanan** [comment]
the cuda variant of the operator gives the correct results.

**Comment by XiaobingSuper** [comment]
The issue wil be solved at next onednn upgradation, thanks!

**Comment by jay746** [comment]
Thanks @XiaobingSuper  and @gchanan for looking into the issue.
I will look forward for the fix in next upgradation.

**Comment by gchanan** [comment]
@XiaobingSuper do you know the surface area of this bug?  It would be helpful to write tests against and potentially workaround until the next onednn upgrade.

**Comment by mingfeima** [comment]
@vpirogov please take a look!

**Comment by vpirogov** [comment]
@gchanan,

This defect impacts weight gradient computations under the following conditions:
* groups = channels (depthwise case)
* stride > 1
* padding > 0

The fix [is available](b7ffe4859a17c360849018b8b4c187ddcdb64dcc) in oneDNN v1.8 release branch. I'll tag a patch release shortly.

**Comment by vpirogov** [comment]
[oneDNN v1.8.1](https://github.com/oneapi-src/oneDNN/releases/tag/v1.8.1) is released.

**Comment by jay746** [comment]
@vpirogov ,
I used pip to install torch and used it.
regarding version, then i see that MKL-DNN v1.6.0 is being used(snippet below).

Can you help me to know the steps to update to point pytorch to use oneDNN v1.8.1? Thanks!

Python 3.6.8 (default, Oct  7 2019, 12:59:55) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
"\n"), sep="\n")>>> print(*torch.__config__.show().split("\n"), sep="\n")
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.6.0 (Git Hash 5ef631a030a6f73131c77892041042805a06064f)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,

**Comment by vpirogov** [comment]
Pinging @XiaobingSuper and @mingfeima for help on the question above.

**Comment by XiaobingSuper** [comment]
@jay746 , just run the following commond to use OneDNN v.1.8.1: 
```
cd third_party/ideep/mkl-dnn && git checkout master && git pull && git checkout v1.8.1
```
and then build pytorch again.

**Comment by VitalyFedyunin** [comment]
@XiaobingSuper please submit PR to update OneDNN & Ideep, we need to have it landed before 1.8 branch cut (Feb 4th)

**Comment by XiaobingSuper** [comment]
> @XiaobingSuper please submit PR to update OneDNN & Ideep, we need to have it landed before 1.8 branch cut (Feb 4th)

Yes, we are under doing some test and will upgrade it soon. Thanks!

**Comment by gchanan** [comment]
I'm closing this because the bug has been resolved and there's already a separate PR for upgrading OneDNN.

**Comment by jay746** [comment]
@gchanan , I had an idea that this issue fix targeted for 1.8.0 version, but i see issue reproducible at 1.8.0 version of Torch

I looked into build config and i see Intel(R) MKL-DNN v1.7.0 used while fix is available in v1.8.1
Can you please comment on which release is targeted for this fix?

**Comment by ngimel** [comment]
Cannot reproduce with 1.8 conda package.

**Comment by jay746** [comment]
I see what is an issue here, original issue was not only with group = 24, but also group = 24 * n (n >=1)
i tried 1.8.0 with group = 24 and it worked fine, but it fails with group = 48/96 

As this bug has been closed, Can we reopen this bug or create duplicate it?

code to reproduce:
`import torch
group_val = 96
ifm = torch.ones([1, group_val, 6, 6], dtype=torch.float32)
weights = torch.ones([group_val, 1, 3, 3], dtype=torch.float32)
op = torch.nn.Conv2d(
        in_channels=group_val,
        out_channels=group_val,
        kernel_size=[3,3],
        stride=[2,2],
        padding=[1,1],
        dilation=[1,1],
        groups=group_val,
        bias=False,
        padding_mode='zeros'
    )

op.weight.data = weights
res = op(ifm)
grad_in = torch.ones(res.shape, dtype=torch.float32)
res.backward(grad_in)
print(op.weight.grad)`

**Comment by ngimel** [comment]
Reopening

**Comment by XiaobingSuper** [comment]
@jay746 , OneDNN v1.8.1 is avilable in the pytorch master which can works well for your case.

**Comment by jay746** [comment]
@XiaobingSuper 
I have installed pytorch pip package(1.8.0)
Can you help me steps how to enable v1.8.1 of mil dnn in torch 1.8.0?

**Comment by ngimel** [comment]
Unfortunately, you cannot swap OneDNN version in pytorch binary build v 1.8, you'd need pytorch nightly for this

**Comment by jay746** [comment]
Can pytorch nightly be downloaded by pip install?
Or I need yo build locally?
( I have something running in CI, so for timing prospective I rely on downloaded package)

**Comment by XiaobingSuper** [comment]
@jay746, you can install pytorch nightly by 
```
pip install --pre torch torchvision torchaudio -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
```

**Comment by jay746** [comment]
Hi,
I had an use case where group=224 and similar issue reproduce. Thought of adding info as this is different than earlier reported condition of group%24 =0 failure. 
If you have test to check this issue then please check for group = 224 as well. Thanks!

**Comment by XiaobingSuper** [comment]
@jay746 , I test it for the pytorch master, I can't reproduce it: 
```
(pytorch-master) [xiaobinz@mlt-clx065 test]$ MKLDNN_VERBOSE=1 python 50042.py
torch.Size([1, 232, 3, 3])
tensor([[[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]],


        ...,


        [[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [6., 9., 9.]]]])
```
please note that it can't call in OneDNN path for this case. So which version of pytorch you are use?

**Comment by jay746** [comment]
@XiaobingSuper , I am on torch 1.6.0
result is as follows:
`(vnv_syn_new) $ pip show torch
Name: torch
Version: 1.6.0
Summary: Tensors and Dynamic neural networks in Python with strong GPU acceleration
Home-page: https://pytorch.org/
Author: PyTorch Team
Author-email: packages@pytorch.org
License: BSD-3
Requires: numpy, future
Required-by: torchvision, torchaudio
(vnv_syn_new) $ python
Python 3.6.8 (default, Oct  7 2019, 12:59:55) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> group_val = 232
>>> ifm = torch.ones([1, group_val, 6, 6], dtype=torch.float32)
>>> weights = torch.ones([group_val, 1, 3, 3], dtype=torch.float32)
>>> op = torch.nn.Conv2d(
...         in_channels=group_val,
...         out_channels=group_val,
...         kernel_size=[3,3],
...         stride=[2,2],
...         padding=[1,1],
...         dilation=[1,1],
...         groups=group_val,
...         bias=False,
...         padding_mode='zeros'
...     )
>>> 
>>> op.weight.data = weights
>>> res = op(ifm)
>>> grad_in = torch.ones(res.shape, dtype=torch.float32)
>>> res.backward(grad_in)
>>> print(op.weight.grad)
tensor([[[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]],


        ...,


        [[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]],


        [[[4., 6., 6.],
          [6., 9., 9.],
          [4., 6., 6.]]]])`

**Comment by XiaobingSuper** [comment]
For this release version, one solution is to disable onednn path by :
```
with torch.backends.mkldnn.flags(enabled=False):
    # runing your model
```
Sorry about this accuracy bug.
